File: NeetCode/Two_Pointers/Valid_Palindrome_II/Valid_Palindrome_II.cpp
// Valid_Palindrome_II.cpp
//
// Compile locally with:
// clang++ Valid_Palindrome_II.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    bool validPalindrome(string s)
    {
        int left(0);
        int right = s.size() - 1;
        while (left < right)
        {
            if (s[left] != s[right])
            {
                return palinSkip(s, left) || palinSkip(s, right);
            }
            left++;
            right--;
        }
        return true;
    }
    bool palinSkip(string s, int n)
    {
        int left(0);
        int right = s.size() - 1;
        while (left < right)
        {
            if (left == n)
            {
                left++;
            }
            else if (right == n)
            {
                right--;
            }
            if (s[left] != s[right])
            {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
};

int main()
{
    cout << "Running Valid_Palindrome_II.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: NeetCode/Two_Pointers/Valid_Palindrome_II/valid__palindrome__i_i.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def validPalindrome(self, s: str) -> bool:
        def palinSkip(n):
            left, right = 0, len(s)
            while left < right:
                if left == n:
                    left += 1
                elif right == n:
                    right -= 1
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1
            return True

        l, r = 0, len(s)
        while l < r:
            if s[l] != s[r]:
                return palinSkip(l) or palinSkip(r)
            l += 1
            r -= 1
        return True


if __name__ == "__main__":
    sol = Solution()
    s = "abca"
    print(sol.validPalindrome(s))

    print("Running Solution...")
